power_grid_model_c/power_grid_model/include/power_grid_model/main_core/output.hpp

```diff
@@ -98,351 +98,335 @@ constexpr ResIt produce_output(MainModelState<ComponentContainer> const& state,
-            return transformer_tap_regulator.get_null_sc_output();
+        state, res_it, [&state, &math_output](Component const& component, Idx2D const math_id) {
+            return output_result<Component>(component, state, math_output, math_id);
+        });
+}
+template <std::derived_from<Base> Component, class ComponentContainer, math_output_type MathOutputType,
+          std::forward_iterator ResIt>
+    requires model_component_state_c<MainModelState, ComponentContainer, Component> &&
+             requires(Component const& component, MainModelState<ComponentContainer> const& state,
+                      std::vector<MathOutputType> const& math_output, Idx obj_seq) {
+                 {
+                     output_result<Component>(component, state, math_output, obj_seq)
+                     } -> std::convertible_to<std::iter_value_t<ResIt>>;
+             }
+constexpr ResIt output_result(MainModelState<ComponentContainer> const& state,
+                              std::vector<MathOutputType> const& math_output, ResIt res_it) {
+    return detail::produce_output<Component, Idx>(
+        state, res_it, [&state, &math_output](Component const& component, Idx const obj_seq) {
+            return output_result<Component, ComponentContainer>(component, state, math_output, obj_seq);
+        });
+}
+template <std::derived_from<Base> Component, class ComponentContainer, math_output_type MathOutputType,
+          std::forward_iterator ResIt>
+    requires model_component_state_c<MainModelState, ComponentContainer, Component> &&
+             requires(Component const& component, std::vector<MathOutputType> const& math_output,
+                      Idx2DBranch3 const& math_id) {
+                 {
+                     output_result<Component>(component, math_output, math_id)
+                     } -> std::convertible_to<std::iter_value_t<ResIt>>;
+             }
+constexpr ResIt output_result(MainModelState<ComponentContainer> const& state,
+                              std::vector<MathOutputType> const& math_output, ResIt res_it) {
+    return detail::produce_output<Component, Idx2DBranch3>(
+        state, res_it, [&math_output](Component const& component, Idx2DBranch3 const& math_id) {
+            return output_result<Component>(component, math_output, math_id);
         });
 }
 
+// output source, load_gen, shunt individually
+template <std::same_as<Appliance> Component, class ComponentContainer, math_output_type MathOutputType,
+          std::forward_iterator ResIt>
+    requires model_component_state_c<MainModelState, ComponentContainer, Component>
+constexpr ResIt output_result(MainModelState<ComponentContainer> const& state,
+                              std::vector<MathOutputType> const& math_output, ResIt res_it) {
+    res_it = output_result<Source>(state, math_output, res_it);
+    res_it = output_result<GenericLoadGen>(state, math_output, res_it);
+    res_it = output_result<Shunt>(state, math_output, res_it);
```

Comment:
I wonder if there is a fancier way for this. For example, append to `res_it` by overloading the `+=` operator for `ResIt`, which is, btw, in its own not a very descriptive name.

Reply:
pass-by-value prevents that.

there are 2 potential solutions:

* rewrite as `std::ranges`
  * not yet possible due to our dependency on `boost` which does not implement all constraints required for `ranges` and the fact that AppleClang does not fully implement `views` yet in the XCode version we use
* create something like an `operator|` overload that pipes iterator argument to the next call.
  * i have done this in the past and it worked fine, but i think it's a little bit overkill for what we're trying to do here.

considering this code was just moved within this file without modifications, i consider this out of scope